Feat halo

dhinf opened the pull request:
added some functionality and minor fixes

dhinf committed:
- minor bug fixes, remvoved dead code
- additional methods for stencil operators
- additional stencil point methodes for coordinate calculatioadditional
stencil point methodes for coordinate calculationn

dhinf committed:
Merge branch 'development' of https://github.com/dash-project/dash into feat-halo

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -76,6 +76,62 @@ class StencilPoint : public Dimensional<int16_t, NumDimensions> {
+   * Returns coordinates adjusted by stencil point
+   */
+  template<typename ElementCoordsT>
+  static ElementCoordsT stencil_coords(ElementCoordsT coords,
```
Why not implement the above function `stencil_coords` in terms of the static `stencil_coords` to reduce code duplication?

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -76,6 +76,62 @@ class StencilPoint : public Dimensional<int16_t, NumDimensions> {
+   * Returns coordinates adjusted by stencil point
+   */
+  template<typename ElementCoordsT, typename ViewSpecT>
+  std::pair<ElementCoordsT,bool> stencil_coords_check(
```
Document what `stencil_coords_check` does different from `stencil_coords`

devreal commented on dash/include/dash/halo/Halo.h:
```diff
@@ -76,6 +76,62 @@ class StencilPoint : public Dimensional<int16_t, NumDimensions> {
+   * Returns coordinates adjusted by stencil point
+   */
+  template<typename ElementCoordsT, typename ViewSpecT>
+  std::pair<ElementCoordsT,bool> stencil_coords_check_abort(
```
Does it abort the program? Where?

devreal commented on dash/include/dash/halo/StencilOperator.h:
```diff
@@ -84,6 +84,34 @@ class StencilOperatorInner {
+   * \param coefficient for center
+   * \param op operation to use (e.g. std::plus). default: std::plus
+   */
+  ElementT get_value_at(
```
Document that the value of the stencil is not modified

devreal commented on dash/include/dash/halo/StencilOperator.h:
```diff
@@ -84,6 +84,34 @@ class StencilOperatorInner {
+   */
+  ElementT get_value_at(
+    const ElementCoords_t& coords, ElementT coefficient_center,
+    std::function<ElementT(const ElementT&, const ElementT&)> op =
```
Make the function a template parameter, e.g., `BinaryFunc`, to avoid creating a `std::function` from lambdas

devreal reviewed (changes requested):
Small comments inline, looks good otherwise :+1:

dhinf committed:
- more documentation
- template based default operations for set and get values methods in
stencil operators

dhinf committed:
- added member variable size to iterator

dhinf committed:
- methods to update inner (fast) and boundary elements

dhinf committed:
- name change

dhinf committed:
- source code format

dhinf committed:
- bug fix for operator+ and operator- (Stenciliterator)
- some documentation

dhinf committed:
Merge branch 'development' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
fix naming

dhinf committed:
Merge branch 'development' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
fixed bug in StencilIterator boundary constructor

bertwesarg commented on dash/include/dash/halo/StencilOperator.h:
```diff
@@ -439,22 +663,22 @@ class StencilOperator {
-
+  StencilOffsets_t set_dimension_offsets() {
+    StencilOffsets_t      dim_offs;
+    signed_pattern_size_t offset = 0;
```
My compiler tells me, that this was declared but never referenced

devreal replied:
@dhinf Can this function be removed?

dhinf replied:
no, it is used.

fmoessbauer commented:
There are compiler errors in clang-mpich using minimal build config:

https://4443-45930061-gh.circle-artifacts.com/0/tmp/build-logs/clang_mpich/20180828-142551--uuid-1f79b8a1/clang/Minimal/build.log
```
In file included from /root/project/dash/include/dash/halo/HaloMatrixWrapper.h:8:
/root/project/dash/include/dash/halo/StencilOperator.h:216:20: warning: 'constexpr' non-static member function will not be implicitly 'const' in C++14; add 'const' to avoid a change in behavior [-Wconstexpr-not-const]
    constexpr void operator()(const StencilOffsets_t& stencil_offs,
                   ^
/root/project/dash/include/dash/halo/StencilOperator.h:232:20: warning: 'constexpr' non-static member function will not be implicitly 'const' in C++14; add 'const' to avoid a change in behavior [-Wconstexpr-not-const]
    constexpr void operator()(const StencilOffsets_t& stencil_offs,
                   ^
/root/project/dash/include/dash/halo/StencilOperator.h:216:20: error: constexpr function's return type 'void' is not a literal type
    constexpr void operator()(const StencilOffsets_t& stencil_offs,
                   ^
/root/project/dash/include/dash/halo/StencilOperator.h:232:20: error: constexpr function's return type 'void' is not a literal type
    constexpr void operator()(const StencilOffsets_t& stencil_offs,
```

dhinf commented:
oh, void is a literal in c++14 but not in 11. Why we still use C++11?

dhinf committed:
Merge branch 'development' of https://github.com/dash-project/dash into feat-halo

dhinf committed:
minor changes comments
